docs/architecture/adr-026-rosetta-api.md

```diff
@@ -0,0 +1,136 @@
+# ADR 26: Rosetta API Support
+
+## Changelog
+
+- August 3rd, 2020: Initial Draft.
+
+## Status
+
+- Approved, pending fine tuning the details of implementation.
+
+## Context
+
+We think it'd be greatly valuable to application developers to have the Cosmos SDK
+provide them with out-of-the-box Rosetta API support.
+According to [the project's website](https://www.rosetta-api.org/), Rosetta API is an open
+standard designed to simplify blockchain deployment and interaction. The latest specifications are
+available at [this URL](https://www.rosetta-api.org/docs/Reference.html).
+
+We want to achieve the following objectives:
+
+- Support multiple versions of Cosmos SDK.
+- Support Cosmos Hub.
+- Implement querying of historical data sets.
+
+## Decision
+
+We intend to develop a library that could be extended and used by application
+developers to integrate an in-process Rosetta API-compliant server with the
+application main binaries. We also intend to provide a standalone gateway server
+program that supports a Cosmos SDK's minimum feature set. Such program could
+run alongside the client applications main binaries.
+
+### Implementation
+
+```
+type Server struct {
+    options
+    router
+}
+
+type Options struct {}
+
+func NewServer(opt Options,  rosettaAdapter RosettaAPIAdapter) {
+    return Server{opt, newRouter(rosettaAdapter)}    
+}
+```
+
+Server is the main type that the client application needs to instantiate. It embeds an `Options` type and exposes a `Start()` method:
+
+```
+func (s Server) Start() error {
+    ...
+    s.ListenAndServe(s.options.endpoint, s.router)
+    ...
+}
+```
+
+An internal `router` type exposes the Rosetta API endpoints:
+
+```
+type router struct {
+    muxRouter mux.router
+    rosettaAdapter RosettaAPIAdapter
+}
+
+func newRouter(adapter RosettaAPIAdapter) {
+    router := &Router{
+        mux: mux.NewRouter(),
+        rosettaAdapter: adapter,
+    }
+
+    router.mux.HandleFunc("/blocks/", router.blocksApi)
+    ...
+    ...
+
+    return router
+}
+```
+
+Each API endpoint has an associated handler, e.g.:
+
+```
+func (r router) blocksApi(w http.ResponseWriter, r *http.Request) {
+    // We build the RosettaGetBlockRequest
+    req := ....
+
+    res := r.rosettaAdapter.GetBlocks(req) 
+    ...
+
+    // We write the response. 
+}
+``` 
+
+The router just translates a rest request into the data struct that the interface of an adapter needs to run.
+The RosettaAdapter interface looks like this:
+
+```
+type RosettaAdapterInterface interface {
+    GetBlock(req RosettaGetBlockRequest) RosettaGetBlockResponse
+    ...
+}
+```
+
+And we will provide different implementations of this `adapter` like for 0.38, for 0.39
```

Comment:
```suggestion
We will provide implementations of this `adapter` for each of the Cosmos SDK supported release series.
```

Reply:
In this case we can just create an Adapter of adapter that query different adapters and return the merged response. 

Query -> AdapterOfAdapters -> Adapter1 (cosmos-sdk 0.37) (Hub1)
                                                   -> Adapter2 (cosmos-sdk 0.38) (Hub2)
                                                   -> Adapter3 (cosmos-sdk 0.38) (Hub3)

The implementation maybe we need to offer for that is one for every SDK and a wrapper adapter of adapters to build complex use cases.